### PR TITLE
perf(ngOptions): prevent initial options repaiting

### PR DESCRIPTION
**Pull request by pbr1111**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix.

**What is the current behavior? (You can also link to an open issue here)**
Refers to this issue: [ngOptions slow performance in IE due to option rerendering](https://github.com/angular/angular.js/issues/15801)

**What is the new behavior (if this is a feature change)?**
Now the options are loaded once when we create the `<select ng-options>` directive. Also, if we add an `<option>` element with an ng-if directive, it is not deleted now.

**Does this PR introduce a breaking change?**
No. 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:
None.

### CODE REVIEW

**Comment by Narretz** [comment]
Hi,
thanks for the Pr. As you can see in the Travis logs, there are a lot of test failures. Most of these come from the fact that there's a different writeSelectValue for select multiple, so you'll to make the change to this fn too.

Please also note that we definitely need to support an empty option with ngIf - there are some tests for this, and I assume some of them will fail in the current state.

You can run the unit tests locally with `grunt autotest`. This runs the test every time you change a file.

**Comment by pbr1111** [comment]
Hi, let's see if I can fix it. Thanks for the info!

**Comment by pbr1111** [comment]
I just uploaded the fixed code. Let's see if it pass the tests.

**Comment by pbr1111** [comment]
The code just passed all the tests. This solution avoids double repainting of `<options>` and also makes ng-if work correctly in the empty option.

**Comment by Narretz** [comment]
Thx @pbr1111, I will review this next week. Was there any reason for closing / reopening the issue so often?

**Comment by pbr1111** [comment]
I close and reopen because eslint tests fails several times and I've made a rebase of my branch each time it fails to upload the new commit with the errors solved.

**Comment by Narretz** [comment]
If you push new / changed code, it'll run Travis again, no need to close / reopen the issue.

File: src/ng/directive/ngOptions.js [hunk 1]
```diff
@@ -624,7 +629,7 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
         // the option list in listbox style, i.e. the select is [multiple], or specifies a [size].
         // See https://github.com/angular/angular.js/issues/11314 for more info.
         // This is unfortunately untestable with unit / e2e tests
-        if (option.label !== element.label) {
+        if (option.label !== undefined) {
```
Comment (gkalpak) [inline src/ng/directive/ngOptions.js]: Why this change?

File: src/ng/directive/ngOptions.js [hunk 2]
```diff
@@ -420,12 +420,19 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
...
       // The emptyOption allows the application developer to provide their own custom "empty"
       // option when the viewValue does not match any of the option values.
-      for (var i = 0, children = selectElement.children(), ii = children.length; i < ii; i++) {
-        if (children[i].value === '') {
-          selectCtrl.hasEmptyOption = true;
-          selectCtrl.emptyOption = children.eq(i);
-          break;
-        }
+      var optionsToRemove = [];
+      for (var i = 0, children = selectElement[0].childNodes, ii = children.length; i < ii; i++) {
```
Comment (gkalpak) [inline src/ng/directive/ngOptions.js]: Is the move from jqLite/jQuery APIs to native APIs part of the perf fix or is it just a "generally good idea" type of change?

Comment (pbr1111) [inline src/ng/directive/ngOptions.js]: jqLite/jQuery children() method doesn't return the nodes of nodeType equal to NODE_TYPE_COMMENT. We can use jqLite/jQuery contents() method instead, but I prefer to use native API.

File: src/ng/directive/ngOptions.js (hunk 1 above)
Comment (pbr1111) [inline src/ng/directive/ngOptions.js]: element.label always will be undefined (because we create a new empty option element each time), then it's useless to read element.label value if we know that its value is undefined.

File: src/ng/directive/ngOptions.js [hunk 3]
```diff
@@ -420,12 +420,19 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
...
+          if (!selectCtrl.hasEmptyOption
+              && (children[i].value === '' || children[i].nodeType === NODE_TYPE_COMMENT)) {
```
Comment (gkalpak) [inline src/ng/directive/ngOptions.js]: So, any comment will be interpreted as empty option? This doesn't seem right.

Comment (pbr1111) [inline src/ng/directive/ngOptions.js]: True. I'm doing changes to my commit. I will simplify this whole part and leave it as close to the previous code.

**Comment by pbr1111** [comment]
To modify the commit, is there any way to do it without having to rebase my branch?

**Comment by pbr1111** [comment]
I just modified the previous commit. Now the select element is emptied after it founds (or not) an empty option (the code is the same as before, it works because the options haven't yet been compiled in postLink, my previous solution was totally incorrect). The empty option element is readded if it's provided.

**Comment by Narretz** [comment]
@pbr1111 You can just push a new commit and you / we will squash before merging.

> it works because the options haven't yet been compiled in postLink, my previous solution was totally incorrect

You should include a test that works with the correct version but fails with the incorrect one - it sounds like we are missing a test there

**Comment by pbr1111** [comment]
I added some tests to check both cases. The second test https://github.com/angular/angular.js/pull/15812/commits/2ec5c07a95d1f6a90151048b49e6579e7cf36539 should fail with the current implementation, but it doesn't fail with https://github.com/angular/angular.js/pull/15812/commits/eb3661c6e4d66c76c3cbaed099b21719308dea6a commit.

File: src/ng/directive/ngOptions.js [hunk 4]
```diff
@@ -449,36 +452,38 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
       if (!multiple) {
 
         selectCtrl.writeValue = function writeNgOptionsValue(value) {
-          var selectedOption = options.selectValueMap[selectElement.val()];
-          var option = options.getOptionFromViewValue(value);
+          if (options) {
```
Comment (Narretz) [inline src/ng/directive/ngOptions.js]: Can you change to `if (!options) return`? This will reduce the whitespace changes.

File: test/ng/directive/ngOptionsSpec.js [hunk 5]
```diff
@@ -2641,6 +2658,23 @@ describe('ngOptions', function() {
...
+
+    it('should ignore the comments within the select', function() {
+      scope.values = [
+        {name:'black'},
+        {name:'white'},
+        {name:'red'}
+      ];
+      scope.isBlank = true;
+
+      createSingleSelect('<!-- test comment --><option ng-if="isBlank" value="">blank</option><!-- test comment 2 -->');
```
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Line is too long, please split at around 100 chars

File: test/ng/directive/ngOptionsSpec.js [hunk 6]
```diff
@@ -2641,6 +2658,23 @@ describe('ngOptions', function() {
       dealoc(element);
     }));
 
+
+    it('should ignore the comments within the select', function() {
```
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: There are no expectations regarding the comments you added. Should this test  that all comments that are not part of the empty option are removed?

File: test/ng/directive/ngOptionsSpec.js [hunk 7]
```diff
@@ -779,6 +779,23 @@ describe('ngOptions', function() {
     expect(options[2]).not.toBeMarkedAsSelected();
   });
 
+
+  it('should render the options only one time', function() {
```
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: the **initial** options

File: test/ng/directive/ngOptionsSpec.js [hunk 8]
```diff
@@ -779,6 +779,23 @@ describe('ngOptions', function() {
...
+    scope.values = ['black', 'white', 'red'];
+    createSelect({
+      'ng-model': 'value'
+    });
+
+    element.attr('ng-options', 'value for value in values');
```
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: you should pass the ngoptions expression like you did with ng-model. Or use createSingleSelect which has useful defaults.

File: src/ng/directive/ngOptions.js (hunk 1 above)
Comment (Narretz) [inline src/ng/directive/ngOptions.js]: You should use `isDefined`here

File: src/ng/directive/ngOptions.js [hunk 9]
```diff
@@ -428,6 +428,9 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
         }
       }
 
+      // Remove all the elements inside the select.
+      selectElement.empty();
```
Comment (Narretz) [inline src/ng/directive/ngOptions.js]: Please add something like:
"The empty option will be compiled and rendered before we first generate the options"

**Review by Narretz — CHANGES_REQUESTED** [review]
Looking good so far! Some minor clarifications and changes nedded.  I would like for @petebacondarwin too also take a look because I remember he once wasn't a fan of doing `if (!options) return`

File: test/ng/directive/ngOptionsSpec.js [hunk 10]
```diff
@@ -779,6 +779,23 @@ describe('ngOptions', function() {
...
+    var repaint = 0;
+    element[0].addEventListener('DOMSubtreeModified', function(ev) {
+      repaint++;
+      expect(repaint).toBeLessThan(2);
```
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Firefox 47 failed this expect. Probably because `DOMSubtreeModified`behaves differently in it. Maybe we should use MutationObserver instead for Browsers that support it?

**Review by Narretz — CHANGES_REQUESTED** [review]

File: test/ng/directive/ngOptionsSpec.js [hunk 11]
```diff
@@ -779,6 +779,32 @@ describe('ngOptions', function() {
...
 
+
+  it('should render the initial options only one time', function() {
+    scope.values = ['black', 'white', 'red'];
+    // Insert a select element into the DOM without ng-options to suscribe to DOM change events
+    createSelect({
+      'ng-model': 'value'
+    });
+
+    // Add ngOptions as attribute before recompiling the select element
```
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Thanks for clarifying why you did it like this ;)
However, recompiling is generally a bad idea, and I wouldn't put it in a test. The better option is to add the listener via a custom directive.

**Review by Narretz — CHANGES_REQUESTED** [review]

**Comment by Narretz** [comment]
Unfortunately, Firefox still reports 2 paints ... :(
